presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tablestore;
+
+import com.facebook.presto.spi.SchemaTableName;
+import com.google.common.collect.ImmutableSet;
+import org.apache.commons.lang3.StringUtils;
+
+import javax.annotation.Nonnull;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.tablestore.TablestoreSessionProperties.MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT;
+import static com.facebook.presto.tablestore.TablestoreSessionProperties.MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS;
+import static com.google.common.base.Preconditions.checkState;
+
+public class IndexFirstSet
```

Comment:
could we have a better name? IndexFirstSet is king of misleading

Reply:
renamed to IndexSelectionStrategy